src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/SignedCms.cs

```diff
@@ -470,5 +470,63 @@ public void CheckHash()
         {
             return ref _signedData;
         }
+
+        public void AddCertificate(X509Certificate2 certificate)
+        {
+            int existingLength = _signedData.CertificateSet?.Length ?? 0;
+
+            byte[] rawData = certificate.RawData;
+
+            if (existingLength > 0)
+            {
+                foreach (CertificateChoiceAsn cert in _signedData.CertificateSet)
+                {
+                    if (cert.Certificate.Value.Span.SequenceEqual(rawData))
+                    {
+                        throw new CryptographicException(SR.Cryptography_Cms_CertificateAlreadyInCollection);
+                    }
+                }
+            }
+
+            if (_signedData.CertificateSet == null)
+            {
+                _signedData.CertificateSet = new CertificateChoiceAsn[1];
+            }
+            else
+            {
+                Array.Resize(ref _signedData.CertificateSet, existingLength + 1);
+            }
+
+            _signedData.CertificateSet[existingLength] = new CertificateChoiceAsn
+            {
+                Certificate = certificate.RawData
```

Comment:
Or if certificate.RawData isn't going to be modified from above, you can keep the rawData local where you have it and just use it here.